### PR TITLE
feat: add event recorder for secrets creation and rotation

### LINKED ISSUE #307: generate events for secret creation and rotation
**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
- Generate events for pod to signal secret creation and failures
- We can also use this event recorder to publish events when the rotation reconciler completes updating the content for the pod. 

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Secrets Store CSI Driver version: (use the image tag):
- Kubernetes version: (use `kubectl version`):

### PR DESCRIPTION
**Pull request by aramase**
**What this PR does / why we need it**:
- Adds event recorder to generate events for secret creation failure/ when mounted contents are rotated
- Adds SHA check to determine if updating secret is required
- 

**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #307 

**Special notes for your reviewer**:
```bash
  Normal  MountRotationComplete   8s    csi-secrets-store-rotation  successfully rotated mounted contents for spc default/azure
  Normal  SecretRotationComplete  8s    csi-secrets-store-rotation  successfully rotated K8s secret foosecret
```

### CODE REVIEW

**Comment by aramase** [comment]
/test pull-secrets-store-csi-driver-e2e-azure

(retrying now that AAD failures should be resolved)

**Comment by aramase** [comment]
@ritazh @tam7t @tomhjp PTAL when you get a chance!

File: controllers/secretproviderclasspodstatus_controller.go [hunk 1]
```diff
@@ -308,7 +350,11 @@ func (r *SecretProviderClassPodStatusReconciler) patchSecretWithOwnerRef(ctx con
 		Name:      name,
 	}
 	err := r.Client.Get(ctx, secretKey, secret)
-	if err != nil && !errors.IsNotFound(err) {
+	if apierrors.IsNotFound(err) {
```
Comment (ritazh) [inline controllers/secretproviderclasspodstatus_controller.go:353]: should we check for nil first?

File: pkg/rotation/reconciler.go [hunk 2]
```diff
@@ -44,19 +48,24 @@ import (
...
+
 	"sigs.k8s.io/secrets-store-csi-driver/apis/v1alpha1"
 	secretsStoreClient "sigs.k8s.io/secrets-store-csi-driver/pkg/client/clientset/versioned"
-
-	log "github.com/sirupsen/logrus"
 )
 
 const (
-	permission os.FileMode = 0644
+	permission                      os.FileMode = 0644
```
Comment (ritazh) [inline pkg/rotation/reconciler.go:64]: nit: space

File: pkg/rotation/reconciler.go [hunk 3]
```diff
@@ -389,11 +415,11 @@ func (r *Reconciler) processNextItem() bool {
 	spcps, err := r.store.GetSecretProviderClassPodStatus(key.(string))
 	if err != nil {
 		log.Errorf("failed to get spc pod status, error: %+v", err)
-		r.handleError(err, key)
+		return true
```
Comment (ritazh) [inline pkg/rotation/reconciler.go:418]: what's the reason for this change?

File: controllers/secretproviderclasspodstatus_controller.go (hunk 1 above)
Comment (aramase) [reply controllers/secretproviderclasspodstatus_controller.go]: *logically* `apierrors.IsNotFound` returns false if error is nil or error is anything other than not found. So we can do the check even without the nil check first.

But I think the code might be more readable with the `err != nil` check. So I've updated the PR.

File: pkg/rotation/reconciler.go (hunk 3 above)
Comment (aramase) [reply pkg/rotation/reconciler.go:451]: Did some research and updated to use rate limited queueing to still allow retries when the `SecretProviderClassPodStatus` is not found. Also added a comment to describe why we are making the changes.

**Review by tam7t — APPROVED** [review]
LGTM

File: pkg/rotation/reconciler.go [hunk 4]
```diff
@@ -286,50 +314,65 @@ func (r *Reconciler) reconcile(ctx context.Context, spcps *v1alpha1.SecretProvid
 		}, updateFn); err != nil {
 			return fmt.Errorf("failed to update spc pod status, err: %+v", err)
```
Comment (ritazh) [inline pkg/rotation/reconciler.go:315]: Do we also want to generate an event for this failure too?

Comment (aramase) [reply pkg/rotation/reconciler.go:316]: Yeah, added an event for failing to spcps with the latest versions.
